src/net.cpp

```diff
@@ -912,6 +912,351 @@ size_t V1Transport::GetSendMemoryUsage() const noexcept
+        if (m_recv_state != RecvState::APP_READY) m_recv_decode_buffer = {};
+    }
+    return true;
+}
+
+bool V2Transport::ReceivedBytes(Span<const uint8_t>& msg_bytes) noexcept
+{
+    AssertLockNotHeld(m_recv_mutex);
+    LOCK(m_recv_mutex);
+    while (!msg_bytes.empty()) {
+        // Decide how many bytes to copy from msg_bytes to m_recv_buffer.
+        size_t max_read = GetMaxBytesToProcess();
+        // Can't read more than provided input.
+        max_read = std::min<uint32_t>(msg_bytes.size(), max_read);
+        // Copy data to buffer.
+        m_recv_buffer.insert(m_recv_buffer.end(), UCharCast(msg_bytes.data()), UCharCast(msg_bytes.data() + max_read));
+        msg_bytes = msg_bytes.subspan(max_read);
+
+        // Process data in the buffer.
+        switch (m_recv_state) {
+        case RecvState::WAITING:
+            if (!m_recv_buffer.empty()) {
+                m_recv_state = RecvState::KEY;
+                LOCK(m_send_mutex);
+                assert(m_send_state == SendState::KEY_WAITING);
+                m_send_state = SendState::KEY;
+            }
+            break;
+
+        case RecvState::KEY:
+            ProcessReceivedKey();
+            break;
+
+        case RecvState::GARB_GARBTERM:
+            if (!ProcessReceivedGarbage()) return false;
+            break;
+
+        case RecvState::GARBAUTH:
+        case RecvState::VERSION:
+        case RecvState::APP:
+            if (!ProcessReceivedPacket()) return false;
+            break;
+
+        case RecvState::APP_READY:
+            return true;
+        }
+    }
+
+    return true;
+}
+
+std::optional<std::string> V2Transport::GetMessageType(Span<const uint8_t>& contents) noexcept
+{
+    if (contents.size() < 13 || contents[0] != 0) return std::nullopt; // Short encoding not yet supported
+
+    size_t msg_type_len{0};
+    while (msg_type_len < 12 && contents[1 + msg_type_len] != 0) {
+        if (contents[1 + msg_type_len] < ' ' || contents[1 + msg_type_len] > 0x7F) {
+            return {};
+        }
+        ++msg_type_len;
+    }
+    std::string ret{reinterpret_cast<const char*>(contents.data() + 1), msg_type_len};
+    while (msg_type_len < 12) {
+        if (contents[1 + msg_type_len] != 0) return {};
+        ++msg_type_len;
+    }
+    contents = contents.subspan(13);
+    return {std::move(ret)};
+}
+
+CNetMessage V2Transport::GetReceivedMessage(std::chrono::microseconds time, bool& reject_message) noexcept
+{
+    AssertLockNotHeld(m_recv_mutex);
+    LOCK(m_recv_mutex);
+    assert(m_recv_state == RecvState::APP_READY);
+    Span<const uint8_t> contents{m_recv_decode_buffer};
+    auto msg_type = GetMessageType(contents);
+    CDataStream ret(m_recv_type, m_recv_version);
+    CNetMessage msg{std::move(ret)};
+    msg.m_raw_message_size = m_recv_decode_buffer.size() + BIP324Cipher::EXPANSION;
+    if (msg_type) {
+        reject_message = false;
+        msg.m_type = std::move(*msg_type);
+        msg.m_time = time;
```

Comment:
In `V1Transport::GetReceivedMessage`, we call `RandAddEvent()` to harvest entropy from the time and checksum of the received messages. Should something similar be done for V2?